js-packages/@prestojs/rest/src/ViewModelEndpoint.ts

```diff
@@ -0,0 +1,111 @@
+import { UrlPattern } from '@prestojs/routing';
+import { isViewModelClass, ViewModelConstructor } from '@prestojs/viewmodel';
+import cloneDeep from 'lodash/cloneDeep';
+import get from 'lodash/get';
+import set from 'lodash/set';
+
+import Endpoint, { EndpointOptions } from './Endpoint';
+
+function cacheDataForModel<T extends ViewModelConstructor<any>>(model: T, data): T | T[] {
+    if (Array.isArray(data)) {
+        const records = data.map(datum => new model(datum));
+        model.cache.addList(records);
+        return records;
+    }
+    const record = new model(data);
+    model.cache.add(record);
+    return record;
+}
+
+type ViewModelMapping = ViewModelConstructor<any> | Record<string, ViewModelConstructor<any>>;
+type ViewModelMappingDef = ViewModelMapping | (() => ViewModelMapping | Promise<ViewModelMapping>);
+
+/**
+ * Define an endpoint for a ViewModel. The response is transformed according
+ * to `viewModelMapping`. If the endpoint returns a single record or a list
+ * of the same record type you can pass a ViewModel class as the parameter:
```

Comment:
This was a bit confusing, I've tried to rewrite it to make it clearer. Feel free to fix/alter anything that I've got wrong/made less clear (in particular, the fact that arrays vs individual objects is handled automatically)

```
/**
 * Define an endpoint for a ViewModel. The response is transformed according
 * to `viewModelMapping`.
 *
 * If the response is an object mapping different models you can specify how
 * each key is transformed:
 *
 * ```js
 * new ViewModelEndpoint(new UrlPattern('/users/'), {
 *     users: User,
 *     bookings: Booking,
 * });
 * ```
 *
 * Dot notation is also supported for nested objects:
 *
 * ```js
 * new ViewModelEndpoint(new UrlPattern('/users/'), {
 *     "records.users": User,
 *     "records.bookings": Booking,
 * });
 * ```
 *
 * If an element is an array then it will transparently be treated as a list of
 * objects of that type and the transformation function will be invoked for each
 * element one by one:
 * 
 * ```js
 * // Response is a single User instance:
 * const userRetrieve = new ViewModelEndpoint(new UrlPattern('/users/:id/'), User);
 * // Response is an array of User instances
 * // (declaration is the same but the response handler will treat it differently)
 * const userList = new ViewModelEndpoint(new UrlPattern('/users/'), User);
 * ```
 *
 * The mapping can optionally be a function that returns a mapping or returns a promise that
 * resolves to a mapping. This is useful for dealing with dynamic imports or returning a class
 * that isn't yet defined.
 *
 * ```js
 * new ViewModelEndpoint(new UrlPattern('/users/'), () => {
 *   const Booking = (await import('./Booking')).default;
 *   return {
 *     "records.users": User,
 *     "records.bookings": Booking,
 *   }
 * });
 * ```
 *
 * If the endpoint is paginated use [PaginatedViewModelEndpoint](doc:PaginatedViewModelEndpoint).
 *
 * TODO: Add note about how foreign key data is cached once we have worked that out
 * TODO: Convention for deleting an item?
 *
 * @extract-docs
 */
```

Reply:
To clarify: if I have an object that contains a nested child object, the parent object transformation function still gets passed the full unabridged data for that object (including the raw data for the child object)?

Reply:
To clarify: the Promises form is only for defining the mapping? If I had a mapping function that returned a Promise that resolves to the mapped data then that wouldn't work? (I don't think it should, but when I initially saw that you could use Promises that's what I assumed it would be for)

Reply:
> To clarify: if I have an object that contains a nested child object, the parent object transformation function still gets passed the full unabridged data for that object (including the raw data for the child object)?

Correct. Typically though I don't think this would be useful - you'd use foreign key fields to handle those automatically for you

> To clarify: the Promises form is only for defining the mapping? If I had a mapping function that returned a Promise that resolves to the mapped data then that wouldn't work? (I don't think it should, but when I initially saw that you could use Promises that's what I assumed it would be for)

Yes that's correct in terms of the intention here but technically both will work as this transformation functions runs as part of a promise chain.. so returning a promise from the function itself will still work (can't think of a case when you'd actually need that though)

Reply:
Thanks - I've updated this now